File: shared/src/commonMain/kotlin/org/nihongo/mochi/ui/settings/SettingsScreen.kt
package org.nihongo.mochi.ui.settings

import androidx.compose.foundation.Image
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import org.jetbrains.compose.resources.DrawableResource
import org.jetbrains.compose.resources.painterResource
import org.jetbrains.compose.resources.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.nihongo.mochi.presentation.MochiBackground
import org.nihongo.mochi.presentation.settings.SettingsViewModel
import org.nihongo.mochi.shared.generated.resources.*
import org.nihongo.mochi.ui.ResourceUtils
import org.nihongo.mochi.domain.services.VoiceGender

// Data class for language items, now used in Compose
data class LanguageItem(val code: String, val name: String, val flagRes: DrawableResource)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SettingsScreen(
    viewModel: SettingsViewModel,
    onThemeChanged: (Boolean) -> Unit,
    onLocaleChanged: (String) -> Unit
) {
    val uiState by viewModel.uiState.collectAsState()
    val scrollState = rememberScrollState()
    
    val languages = remember {
        listOf(
            LanguageItem("ar_SA", "العربية", Res.drawable.flag_sa_sa),
            LanguageItem("bn_BD", "বাংলা", Res.drawable.flag_bn),
            LanguageItem("de_DE", "Deutsch", Res.drawable.flag_de),
            LanguageItem("en_GB", "English", Res.drawable.flag_en_gb),
            LanguageItem("es_ES", "Español", Res.drawable.flag_es),
            LanguageItem("fr_FR", "Français", Res.drawable.flag_fr_fr),
            LanguageItem("in_ID", "Bahasa Indonesia", Res.drawable.flag_id),
            LanguageItem("it_IT", "Italiano", Res.drawable.flag_it),
            LanguageItem("ja_JP", "日本語", Res.drawable.flag_jp),
            LanguageItem("ko_KR", "한국어", Res.drawable.flag_kr),
            LanguageItem("mn_MN", "Монгол", Res.drawable.flag_mn),
            LanguageItem("pt_BR", "Português", Res.drawable.flag_pt_br),
            LanguageItem("ru_RU", "Русский", Res.drawable.flag_ru),
            LanguageItem("th_TH", "ไทย", Res.drawable.flag_th_th),
            LanguageItem("ua_UA", "Українська", Res.drawable.flag_ua),
            LanguageItem("vi_VN", "Tiếng Việt", Res.drawable.flag_vn),
            LanguageItem("zh_CN", "简体中文", Res.drawable.flag_cn)
        )
    }

    val learningModes = uiState.availableModes.ifEmpty { 
        listOf("JLPT", "School", "Challenge") 
    }

    MochiBackground {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(scrollState)
                .padding(16.dp)
        ) {
            Text(
                text = stringResource(Res.string.settings_title),
                fontSize = 34.sp,
                fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.primary,
                modifier = Modifier.padding(bottom = 24.dp)
            )

            SettingsSection(title = stringResource(Res.string.settings_category_general)) {
                // Language Spinner
                var expanded by remember { mutableStateOf(false) }
                val selectedLanguage = languages.find { it.code == uiState.currentLocaleCode } ?: languages.first()

                ExposedDropdownMenuBox(
                    expanded = expanded,
                    onExpandedChange = { expanded = !expanded }
                ) {
                    OutlinedTextField(
                        value = selectedLanguage.name,
                        onValueChange = {},
                        readOnly = true,
                        label = { Text(stringResource(Res.string.settings_language)) },
                        leadingIcon = {
                            Image(painter = painterResource(selectedLanguage.flagRes), contentDescription = null, modifier = Modifier.size(24.dp))
                        },
                        trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
                        modifier = Modifier
                            .fillMaxWidth()
                            .menuAnchor()
                    )

                    ExposedDropdownMenu(
                        expanded = expanded,
                        onDismissRequest = { expanded = false }
                    ) {
                        languages.forEach { language ->
                            DropdownMenuItem(
                                text = { 
                                    Row(verticalAlignment = Alignment.CenterVertically) {
                                        Image(painter = painterResource(language.flagRes), contentDescription = null, modifier = Modifier.size(24.dp))
                                        Spacer(Modifier.width(8.dp))
                                        Text(language.name)
                                    }
                                },
                                onClick = {
                                    viewModel.onLocaleChanged(language.code)
                                    onLocaleChanged(language.code)
                                    expanded = false
                                }
                            )
                        }
                    }
                }

                Spacer(Modifier.height(16.dp))
                
                // Pronunciation
                Text(
                    text = stringResource(Res.string.settings_pronunciation),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onSurface
                )
                Column {
                    Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.clickable { viewModel.onPronunciationChanged("Roman") }) {
                        RadioButton(selected = uiState.pronunciation == "Roman", onClick = { viewModel.onPronunciationChanged("Roman") })
                        Text(stringResource(Res.string.settings_pronunciation_roman), color = MaterialTheme.colorScheme.onSurface)
                    }
                    Spacer(Modifier.height(8.dp))
                    Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.clickable { viewModel.onPronunciationChanged("Hiragana") }) {
                        RadioButton(selected = uiState.pronunciation == "Hiragana", onClick = { viewModel.onPronunciationChanged("Hiragana") })
                        Text(stringResource(Res.string.settings_pronunciation_hiragana), color = MaterialTheme.colorScheme.onSurface)
                    }
                }
            }
            
            Spacer(Modifier.height(16.dp))
            
            // Text-to-Speech Section
            SettingsSection(title = stringResource(Res.string.settings_tts_category)) {
                // TTS Speed
                Text(text = stringResource(Res.string.settings_tts_speed), color = MaterialTheme.colorScheme.onSurface)
                SliderWithLabel(
                    value = uiState.ttsRate,
                    onValueChange = { viewModel.onTtsRateChanged(it) },
                    valueRange = 0.5f..2.0f
                )
                
                Spacer(Modifier.height(16.dp))

                // TTS Voice Selection
                Text(text = stringResource(Res.string.settings_tts_voice_selection), color = MaterialTheme.colorScheme.onSurface)
                var expandedVoices by remember { mutableStateOf(false) }
                
                val isArabic = uiState.currentLocaleCode.startsWith("ar")
                
                // Helper to get friendly name
                @Composable
                fun getFriendlyName(voiceId: String?): String {
                    return when {
                        voiceId == null -> stringResource(Res.string.settings_tts_voice_default)
                        voiceId.contains("jad-local") -> stringResource(Res.string.settings_tts_voice_male)
                        voiceId.contains("jab-local") -> stringResource(Res.string.settings_tts_voice_female)
                        else -> voiceId
                    }
                }

                val selectedVoiceLabel = getFriendlyName(uiState.selectedVoiceId)

                Row(verticalAlignment = Alignment.CenterVertically) {
                    ExposedDropdownMenuBox(
                        expanded = expandedVoices,
                        onExpandedChange = { if (!isArabic) expandedVoices = !expandedVoices },
                        modifier = Modifier.weight(1f)
                    ) {
                        OutlinedTextField(
                            value = if (isArabic) stringResource(Res.string.settings_tts_voice_forced_arabic) else selectedVoiceLabel,
                            onValueChange = {},
                            readOnly = true,
                            enabled = !isArabic,
                            label = { Text(stringResource(Res.string.settings_tts_available_voices)) },
                            trailingIcon = { if (!isArabic) ExposedDropdownMenuDefaults.TrailingIcon(expanded = expandedVoices) },
                            modifier = Modifier.fillMaxWidth().menuAnchor()
                        )

                        if (!isArabic) {
                            ExposedDropdownMenu(expanded = expandedVoices, onDismissRequest = { expandedVoices = false }) {
                                DropdownMenuItem(
                                    text = { Text(stringResource(Res.string.settings_tts_voice_default)) },
                                    onClick = {
                                        viewModel.onTtsVoiceSelected(null)
                                        expandedVoices = false
                                    }
                                )
                                uiState.availableVoices.forEach { voiceId ->
                                    DropdownMenuItem(
                                        text = { Text(getFriendlyName(voiceId)) },
                                        onClick = {
                                            viewModel.onTtsVoiceSelected(voiceId)
                                            expandedVoices = false
                                        }
                                    )
                                }
                            }
                        }
                    }
                    
                    Spacer(Modifier.width(8.dp))
                    
                    FilledIconButton(
                        onClick = { viewModel.testSpeak() },
                        modifier = Modifier.size(48.dp)
                    ) {
                        Icon(Icons.Default.PlayArrow, contentDescription = "Test Voice")
                    }
                }
            }

            Spacer(Modifier.height(16.dp))
            
            // Learning Mode Section
            SettingsSection(title = stringResource(Res.string.settings_learning_mode)) {
                var expanded by remember { mutableStateOf(false) }
                val rawSelectedMode = if (learningModes.contains(uiState.currentMode)) uiState.currentMode else learningModes.firstOrNull() ?: ""

                @Composable
                fun getModeLabel(mode: String): String {
                    val key = "section_" + mode.lowercase()
                    val resource = ResourceUtils.resolveStringResource(key) ?: ResourceUtils.resolveStringResource(mode.lowercase())
                    return if (resource != null) stringResource(resource) else mode.replaceFirstChar { it.titlecase() }
                }

                ExposedDropdownMenuBox(
                    expanded = expanded,
                    onExpandedChange = { expanded = !expanded }
                ) {
                    OutlinedTextField(
                        value = getModeLabel(rawSelectedMode),
                        onValueChange = {},
                        readOnly = true,
                        label = { Text("Mode") },
                        trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
                        modifier = Modifier.fillMaxWidth().menuAnchor()
                    )

                    ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
                        learningModes.forEach { mode ->
                            DropdownMenuItem(
                                text = { Text(getModeLabel(mode)) },
                                onClick = {
                                    viewModel.onModeChanged(mode)
                                    expanded = false
                                }
                            )
                        }
                    }
                }
            }

            Spacer(Modifier.height(16.dp))

            SettingsSection(title = stringResource(Res.string.settings_category_learning)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Checkbox(checked = uiState.addWrongAnswers, onCheckedChange = { viewModel.onAddWrongAnswersChanged(it) })
                    Text(stringResource(Res.string.settings_add_wrong_answers), color = MaterialTheme.colorScheme.onSurface)
                }
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Checkbox(checked = uiState.removeGoodAnswers, onCheckedChange = { viewModel.onRemoveGoodAnswersChanged(it) })
                    Text(stringResource(Res.string.settings_remove_good_answers), color = MaterialTheme.colorScheme.onSurface)
                }
            }
            
            Spacer(Modifier.height(16.dp))

            SettingsSection(title = stringResource(Res.string.settings_category_interface)) {
                Text(stringResource(Res.string.settings_text_size), color = MaterialTheme.colorScheme.onSurface)
                SliderWithLabel(value = uiState.textSize, onValueChange = { viewModel.onTextSizeChanged(it) })
                
                Spacer(Modifier.height(8.dp))

                Text(stringResource(Res.string.settings_animation_speed), color = MaterialTheme.colorScheme.onSurface)
                SliderWithLabel(value = uiState.animationSpeed, onValueChange = { viewModel.onAnimationSpeedChanged(it) })
                
                Spacer(Modifier.height(8.dp))
                
                Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
                    Text(stringResource(Res.string.settings_theme), color = MaterialTheme.colorScheme.onSurface)
                    Spacer(Modifier.weight(1f))
                    Switch(checked = uiState.isDarkMode, onCheckedChange = {
                        viewModel.onThemeChanged(it)
                        onThemeChanged(it)
                    })
                }
            }
        }
    }
}

@Composable
fun SettingsSection(title: String, content: @Composable () -> Unit) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface.copy(alpha = 0.9f)),
        elevation = CardDefaults.cardElevation(2.dp)
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Text(
                text = title,
                fontSize = 20.sp,
                fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.onSurface,
                modifier = Modifier.padding(bottom = 16.dp)
            )
            content()
        }
    }
}

@Composable
fun SliderWithLabel(
    value: Float, 
    onValueChange: (Float) -> Unit,
    valueRange: ClosedFloatingPointRange<Float> = 0.1f..4.0f
) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        Slider(
            value = value,
            onValueChange = onValueChange,
            valueRange = valueRange,
            modifier = Modifier.weight(1f)
        )
        Spacer(Modifier.width(8.dp))
        val formattedValue = (Math.round(value * 10) / 10.0).toString() + "x"
        Text(
            text = formattedValue,
            modifier = Modifier.widthIn(min = 40.dp),
            color = MaterialTheme.colorScheme.onSurface
        )
    }
}

File: flutter_app/lib/settings_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'providers/settings_provider.dart';
import 'widgets/mochi_background.dart';

class LanguageItem {
  final String code;
  final String name;
  final String flagAsset;

  LanguageItem(this.code, this.name, this.flagAsset);
}

class SettingsScreen extends StatelessWidget {
  const SettingsScreen({super.key});

  @override
  Widget build(BuildContext context) {
    final settings = context.watch<SettingsProvider>();

    final languages = [
      LanguageItem("ar_SA", "العربية", "assets/drawable/flag_sa_sa.png"),
      LanguageItem("bn_BD", "বাংলা", "assets/drawable/flag_bn.png"),
      LanguageItem("de_DE", "Deutsch", "assets/drawable/flag_de.png"),
      LanguageItem("en_GB", "English", "assets/drawable/flag_en_gb.png"),
      LanguageItem("es_ES", "Español", "assets/drawable/flag_es.png"),
      LanguageItem("fr_FR", "Français", "assets/drawable/flag_fr_fr.png"),
      LanguageItem("in_ID", "Bahasa Indonesia", "assets/drawable/flag_id.png"),
      LanguageItem("it_IT", "Italiano", "assets/drawable/flag_it.png"),
      LanguageItem("ja_JP", "日本語", "assets/drawable/flag_jp.png"),
      LanguageItem("ko_KR", "한국어", "assets/drawable/flag_kr.png"),
      LanguageItem("mn_MN", "Монгол", "assets/drawable/flag_mn.png"),
      LanguageItem("pt_BR", "Português", "assets/drawable/flag_pt_br.png"),
      LanguageItem("ru_RU", "Русский", "assets/drawable/flag_ru.png"),
      LanguageItem("th_TH", "ไทย", "assets/drawable/flag_th_th.png"),
      LanguageItem("ua_UA", "Українська", "assets/drawable/flag_ua.png"),
      LanguageItem("vi_VN", "Tiếng Việt", "assets/drawable/flag_vn.png"),
      LanguageItem("zh_CN", "简体中文", "assets/drawable/flag_cn.png"),
    ];

    final selectedLanguage = languages.firstWhere(
      (l) => l.code == settings.currentLocaleCode,
      orElse: () => languages.firstWhere((l) => l.code == "en_GB", orElse: () => languages.first),
    );

    return Scaffold(
      appBar: AppBar(
        title: Text(settings.getString("settings_title"), style: TextStyle(color: Theme.of(context).colorScheme.onBackground)),
        backgroundColor: Colors.transparent,
        elevation: 0,
        iconTheme: IconThemeData(color: Theme.of(context).colorScheme.onBackground),
      ),
      extendBodyBehindAppBar: true,
      body: MochiBackground(
        child: SafeArea(
          child: SingleChildScrollView(
            padding: const EdgeInsets.all(16.0),
            child: Column(
              children: [
                _buildSection(
                  context,
                  title: settings.getString("settings_category_interface"),
                  children: [
                    SwitchListTile(
                      title: Text(settings.getString("settings_theme")),
                      value: settings.isDarkMode,
                      onChanged: (val) => settings.toggleTheme(val),
                    ),
                    const Divider(),
                    ListTile(
                      title: Text(settings.getString("settings_language")),
                      trailing: Row(
                        mainAxisSize: MainAxisSize.min,
                        children: [
                          Image.asset(selectedLanguage.flagAsset, width: 24),
                          const SizedBox(width: 8),
                          Text(selectedLanguage.name),
                          const Icon(Icons.arrow_drop_down),
                        ],
                      ),
                      onTap: () => _showLanguagePicker(context, settings, languages),
                    ),
                  ],
                ),
                const SizedBox(height: 16),
                _buildSection(
                  context,
                  title: settings.getString("settings_category_general"),
                  children: [
                    _buildSliderTile(
                      context,
                      title: settings.getString("settings_animation_speed"),
                      value: settings.animationSpeed,
                      onChanged: (val) => settings.updateAnimationSpeed(val),
                      min: 0.5,
                      max: 2.0,
                    ),
                    const Divider(),
                    _buildSliderTile(
                      context,
                      title: settings.getString("settings_audio_volume"),
                      value: settings.audioVolume,
                      onChanged: (val) => settings.updateAudioVolume(val),
                      min: 0.0,
                      max: 1.0,
                    ),
                    const Divider(),
                    _buildSliderTile(
                      context,
                      title: settings.getString("settings_tts_speed"),
                      value: settings.ttsRate,
                      onChanged: (val) => settings.updateTtsRate(val),
                      min: 0.5,
                      max: 2.0,
                    ),
                    const Divider(),
                    ListTile(
                      title: Text(settings.getString("settings_tts_voice_selection")),
                      trailing: Row(
                        mainAxisSize: MainAxisSize.min,
                        children: [
                          DropdownButton<String?>(
                            value: settings.selectedVoiceId,
                            items: [
                              DropdownMenuItem<String?>(
                                value: null,
                                child: Text(settings.getString("settings_tts_voice_default")),
                              ),
                              ...settings.availableVoices.map((voiceId) => DropdownMenuItem<String?>(
                                    value: voiceId,
                                    child: Text(_getFriendlyVoiceName(voiceId, settings)),
                                  )),
                            ],
                            onChanged: settings.currentLocaleCode.startsWith("ar") 
                                ? null 
                                : (val) => settings.updateTtsVoiceId(val),
                          ),
                          const SizedBox(width: 8),
                          IconButton(
                            icon: const Icon(Icons.play_arrow),
                            onPressed: () => settings.testSpeak(),
                          ),
                        ],
                      ),
                    ),
                  ],
                ),
                const SizedBox(height: 16),
                _buildSection(
                  context,
                  title: settings.getString("settings_category_learning"),
                  children: [
                    ListTile(
                      title: Text(settings.getString("settings_learning_mode")),
                      trailing: DropdownButton<String>(
                        value: settings.currentMode,
                        items: ["JLPT", "School", "Challenge"]
                            .map((m) => DropdownMenuItem(
                                  value: m,
                                  child: Text(m == "JLPT"
                                      ? settings.getString("section_jlpt")
                                      : m == "School"
                                          ? settings.getString("section_school")
                                          : settings.getString("section_challenges")),
                                ))
                            .toList(),
                        onChanged: (val) => settings.updateMode(val!),
                      ),
                    ),
                    const Divider(),
                    ListTile(
                      title: Text(settings.getString("settings_pronunciation")),
                      trailing: DropdownButton<String>(
                        value: settings.pronunciation,
                        items: ["Roman", "Hiragana"]
                            .map((p) => DropdownMenuItem(
                                  value: p,
                                  child: Text(p == "Roman"
                                      ? settings.getString("settings_pronunciation_roman")
                                      : settings.getString("settings_pronunciation_hiragana")),
                                ))
                            .toList(),
                        onChanged: (val) => settings.updatePronunciation(val!),
                      ),
                    ),
                    const Divider(),
                    CheckboxListTile(
                      title: Text(settings.getString("settings_add_wrong_answers")),
                      value: settings.addWrongAnswers,
                      onChanged: (val) => settings.toggleAddWrongAnswers(val!),
                    ),
                    CheckboxListTile(
                      title: Text(settings.getString("settings_remove_good_answers")),
                      value: settings.removeGoodAnswers,
                      onChanged: (val) => settings.toggleRemoveGoodAnswers(val!),
                    ),
                  ],
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  String _getFriendlyVoiceName(String? voiceId, SettingsProvider settings) {
    if (voiceId == null || voiceId.isEmpty) {
      return settings.getString("settings_tts_voice_default");
    }
    final name = voiceId.toLowerCase();
    if (name.contains("jad-local") || name.contains("-m-") || name.contains("male")) {
      return settings.getString("settings_tts_voice_male");
    }
    if (name.contains("jab-local") || name.contains("-f-") || name.contains("female")) {
      return settings.getString("settings_tts_voice_female");
    }
    if (name.contains("sjp-local")) {
      return "${settings.getString("settings_tts_voice_male")} (Samsung)";
    }
    if (name.contains("sja-local")) {
      return "${settings.getString("settings_tts_voice_female")} (Samsung)";
    }
    return voiceId;
  }

  Widget _buildSection(BuildContext context, {required String title, required List<Widget> children}) {
    final theme = Theme.of(context);
    return Card(
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
      color: theme.colorScheme.surface.withOpacity(0.9),
      elevation: 2,
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              title,
              style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold, color: theme.colorScheme.primary),
            ),
            const SizedBox(height: 8),
            ...children,
          ],
        ),
      ),
    );
  }

  Widget _buildSliderTile(
    BuildContext context, {
    required String title,
    required double value,
    required ValueChanged<double> onChanged,
    required double min,
    required double max,
  }) {
    final theme = Theme.of(context);
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(title, style: TextStyle(color: theme.colorScheme.onSurface)),
        Row(
          children: [
            Expanded(
              child: Slider(
                value: value,
                min: min,
                max: max,
                activeColor: theme.colorScheme.primary,
                onChanged: onChanged,
              ),
            ),
            Text("${value.toStringAsFixed(1)}x"),
          ],
        ),
      ],
    );
  }

  void _showLanguagePicker(BuildContext context, SettingsProvider settings, List<LanguageItem> languages) {
    showModalBottomSheet(
      context: context,
      shape: const RoundedRectangleBorder(borderRadius: BorderRadius.vertical(top: Radius.circular(20))),
      builder: (context) => ListView.builder(
        shrinkWrap: true,
        itemCount: languages.length,
        itemBuilder: (context, index) {
          final lang = languages[index];
          return ListTile(
            leading: Image.asset(lang.flagAsset, width: 32),
            title: Text(lang.name),
            onTap: () {
              settings.updateLocale(lang.code);
              Navigator.pop(context);
            },
          );
        },
      ),
    );
  }
}
